orttraining/orttraining/training_api/module.cc

```diff
@@ -26,30 +23,21 @@ const std::string ACCUMULATE_GRAD_CONTROL_INPUT_NAME{"lazy_reset_grad"};
 
 std::unordered_set<const Node*> GetReverseReachableNodes(Graph& inference_graph,
                                                          InlinedVector<const NodeArg*>& output_node_args) {
-  // Perform a bfs from the graph outputs to collect all reachable nodes from the outputs
-  std::queue<const Node*> node_queue;
+  // Perform a graph traversal from the graph outputs to collect all reachable nodes from the outputs
+  InlinedVector<NodeIndex> nodes;
+  nodes.reserve((output_node_args.size()));
   std::unordered_set<const Node*> visited_nodes;
   for (auto node_arg : output_node_args) {
     auto* node = inference_graph.GetProducerNode(node_arg->Name());
     if (!node) {
       continue;
     }
 
-    node_queue.push(node);
+    nodes.push_back(node->Index());
```

Comment:
not sure is it safe we insert a same value of node index twice. maybe better we do a check here?